diff --git a/dwm.c b/dwm.c
index 2661d55..cb65c0d 100644
--- a/dwm.c
+++ b/dwm.c
@@ -186,6 +186,7 @@ static void mappingnotify(XEvent *e);
 static void maprequest(XEvent *e);
 static void monocle(Monitor *m);
 static void tcl(Monitor *m);
+static void ntcl(Monitor *m);
 static void motionnotify(XEvent *e);
 static void movemouse(const Arg *arg);
 static Client *nexttagged(Client *c);
@@ -1234,6 +1235,82 @@ tcl(Monitor * m)
 	}
 }
 
+void
+ntcl(Monitor * m)
+{
+	int x, y, h, w, mw, sw, bdw;
+	unsigned int i, n;
+	Client *c;
+
+	for (n = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), n++) ;
+	if (n == 0)
+		return;
+
+	if (n - MIN(n, m->nmaster) <= 1) {
+		w = m->ww / n;
+		for (i = 0, c = nexttiled(m->clients); c && i < n;
+		     i++, c = nexttiled(c->next)) {
+			resize(c, m->wx + i * w, m->wy, w, m->wh, False);
+		}
+		return;
+	}
+
+	mw = m->mfact * m->ww;
+	sw = (m->ww - mw) / 2;
+
+	for (i = 0, c = nexttiled(m->clients); c && i < MIN(m->nmaster, n);
+	     i++, c = nexttiled(c->next)) {
+		resize(c, m->wx + sw + (i * mw / m->nmaster), m->wy,
+		       mw / m->nmaster, m->wh, False);
+	}
+	bdw = (2 * c->bw);
+
+	n -= MIN(m->nmaster, n);
+
+	if (n <= 0)
+		return;
+
+	w = (m->ww - mw) / 2;
+
+	// right column
+	x = sw + mw;
+	y = m->wy;
+	h = m->wh / (n / 2);
+
+	if (h < bh)
+		h = m->wh;
+
+	for (i = 0; c && i < n / 2; c = nexttiled(c->next), i++) {
+		resize(c,
+		       x,
+		       y, w, (i + 1 == n / 2) ? m->wy + m->wh - y : h, False);
+
+		if (h != m->wh)
+			y = c->y + HEIGHT(c);
+	}
+
+	x = (n + 1 / 2) == 1 ? mw : m->wx;
+	y = m->wy;
+	h = m->wh / ((n + 1) / 2);
+
+	if (h < bh)
+		h = m->wh;
+
+	// left column
+	for (i = 0; c; c = nexttiled(c->next), i++) {
+		resize(c,
+		       x,
+		       y,
+		       (i + 1 == (n + 1) / 2) ? w - bdw : w - bdw,
+		       (i + 1 ==
+			(n + 1) / 2) ? m->wy + m->wh - y - bdw : h - bdw,
+		       False);
+
+		if (h != m->wh)
+			y = c->y + HEIGHT(c);
+	}
+}
+
 void
 motionnotify(XEvent *e)
 {
